### PR TITLE
Add in-memory log caching and a logs dialog

### PR DESCRIPTION
**Pull request by mattprecious**
Don't really like these being in a dialog... but I couldn't come up with another way to add them to the drawer while keeping the light theme.

screenshot-2015-03-25_14 35 34 587

### CODE REVIEW

File: src/debug/java/com/jakewharton/u2020/ui/logs/LogsDialog.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+  private void share() {
+    Intent sendIntent = new Intent();
+    sendIntent.setAction(Intent.ACTION_SEND);
+    sendIntent.putExtra(Intent.EXTRA_TEXT, createShareBody());
+    sendIntent.setType("text/plain");
+    getContext().startActivity(sendIntent);
+  }
+
+  private String createShareBody() {
```
Comment (mattprecious) [inline src/debug/java/com/jakewharton/u2020/ui/logs/LogsDialog.java]: Eh... this isn't very appealing. Will probably change this to a file attachment at some point.

File: src/main/java/com/jakewharton/u2020/data/LumberYard.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Deque;
+import java.util.List;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import rx.Observable;
+import rx.subjects.PublishSubject;
+import timber.log.Timber;
+
+@Singleton
+public class LumberYard {
```
Comment (JakeWharton) [inline src/main/java/com/jakewharton/u2020/data/LumberYard.java]: Yesssssss I'm so glad this name won.

File: src/debug/java/com/jakewharton/u2020/ui/logs/LogAdapter.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  static final class LogItemViewHolder {
+    private final View rootView;
+    @InjectView(R.id.debug_log_level) TextView levelView;
+    @InjectView(R.id.debug_log_tag) TextView tagView;
+    @InjectView(R.id.debug_log_message) TextView messageView;
+
+    public LogItemViewHolder(View rootView) {
+      this.rootView = rootView;
+      ButterKnife.inject(this, rootView);
+      rootView.setTag(this);
```
Comment (JakeWharton) [inline src/debug/java/com/jakewharton/u2020/ui/logs/LogAdapter.java]: I don't like this side-effect. Can you do it in a static method rather than ctor?

Comment (JakeWharton) [inline src/debug/java/com/jakewharton/u2020/ui/logs/LogAdapter.java]: Or just call `setTag` in the `newView` method. The disconnect between extracting this from the tag in `bindView` and not setting it in `newView` is what I don't like.

File: src/main/java/com/jakewharton/u2020/U2020App.java [hunk 4]
```diff
@@ -28,6 +30,7 @@
 
     buildObjectGraphAndInject();
 
+    Timber.plant(lumberYard.tree());
```
Comment (JakeWharton) [inline src/main/java/com/jakewharton/u2020/U2020App.java]: This line of code makes life worth living.

File: src/main/java/com/jakewharton/u2020/data/LumberYard.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+  public List<Entry> bufferedLogs() {
+    return new ArrayList<>(entries);
+  }
+
+  public Observable<Entry> logs() {
+    return entrySubject;
+  }
+
+  public static class Entry {
```
Comment (JakeWharton) [inline src/main/java/com/jakewharton/u2020/data/LumberYard.java]: hahahahahahahahah +final

File: src/debug/java/com/jakewharton/u2020/ui/logs/LogsDialog.java (hunk 1 above)
Comment (JakeWharton) [inline src/debug/java/com/jakewharton/u2020/ui/logs/LogsDialog.java]: :+1: sgtm

File: src/debug/java/com/jakewharton/u2020/ui/logs/LogsDialog.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @Override protected void onStart() {
+    super.onStart();
+
+    adapter.setLogs(lumberYard.bufferedLogs());
+
+    subscriptions = new CompositeSubscription();
+    subscriptions.add(lumberYard.logs() //
+        .subscribeOn(AndroidSchedulers.mainThread()) //
```
Comment (JakeWharton) [inline src/debug/java/com/jakewharton/u2020/ui/logs/LogsDialog.java]: I think you want `observeOn` here... although the effect is the same.

File: src/debug/java/com/jakewharton/u2020/ui/logs/LogAdapter.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    INFO("I", R.color.debug_log_accent_info),
+    WARN("W", R.color.debug_log_accent_warn),
+    ERROR("E", R.color.debug_log_accent_error),
+    ASSERT("A", R.color.debug_log_accent_error),
+    UNKNOWN("?", R.color.debug_log_accent_unknown);
+
+    public final String letter;
+    public final int backgroundColorResId;
+
+    LogLevelStyle(String letter, int backgroundColorResId) {
```
Comment (JakeWharton) [inline src/debug/java/com/jakewharton/u2020/ui/logs/LogAdapter.java]: `@ColorRes` on second param!

**Comment by JakeWharton** [comment]
lgtm :shipit:

File: src/main/java/com/jakewharton/u2020/data/LumberYard.java (hunk 2 above)
Comment (mattprecious) [inline src/main/java/com/jakewharton/u2020/data/LumberYard.java]: Me too!

File: src/debug/java/com/jakewharton/u2020/ui/logs/LogAdapter.java (hunk 3 above)
Comment (mattprecious) [inline src/debug/java/com/jakewharton/u2020/ui/logs/LogAdapter.java]: Moved to `newView`.

File: src/main/java/com/jakewharton/u2020/U2020App.java (hunk 4 above)
Comment (mattprecious) [inline src/main/java/com/jakewharton/u2020/U2020App.java]: I've got plenty more names just waiting for reasons to use them.

File: src/main/java/com/jakewharton/u2020/data/LumberYard.java (hunk 5 above)
Comment (mattprecious) [inline src/main/java/com/jakewharton/u2020/data/LumberYard.java]: Ugh! I thought I got them all this time.

File: src/debug/java/com/jakewharton/u2020/ui/logs/LogsDialog.java (hunk 6 above)
Comment (mattprecious) [inline src/debug/java/com/jakewharton/u2020/ui/logs/LogsDialog.java]: Whoops, I do.

**Pull request merged**